librad/src/meta/project.rs

```diff
@@ -87,6 +154,153 @@ impl Project {
+        })
+    }
+
+    pub fn add_signature(&mut self, key: &Key) -> Result<(), Error> {
+        let id = PeerId::from(key.clone());
+        if self.signatures.iter().any(|s| s.key == id) {
+            Err(Error::SignatureAlreadyPresent)
+        } else if !self.maintainers.iter().any(|m| m == &id) {
+            Err(Error::SignatureFromNonMaintainer)
+        } else {
+            self.signatures.push(self.build_signature(key)?);
+            Ok(())
+        }
+    }
+
+    pub fn remove_signature(&mut self, id: &PeerId) {
+        self.signatures.retain(|s| s.key != *id);
+    }
+
+    pub fn verify_signature(&self, signature: &ProjectSignature) -> Result<(), Error> {
+        signature.verify_data(&self.canonical_data()?)
+    }
+
+    pub fn verify_signatures(&self) -> Result<(), Error> {
+        let data = self.canonical_data()?;
+        for s in self.signatures.iter() {
+            s.verify_data(&data)?
+        }
+        Ok(())
+    }
+
+    pub fn check_signatures_against_maintainers(&self) -> Result<(), Error> {
+        let mut maintainers = HashSet::<&PeerId>::new();
+        for m in self.maintainers.iter() {
+            maintainers.insert(m);
+        }
+
+        for s in self.signatures.iter() {
+            let m = maintainers.take(&s.key);
+            match m {
+                Some(_) => {}
+                None => return Err(Error::SignatureFromNonMaintainer),
+            }
+        }
+
+        if !maintainers.is_empty() {
+            return Err(Error::SignatureMissing);
+        }
+
+        Ok(())
+    }
+
+    pub fn check_validity(&self) -> Result<(), Error> {
+        self.check_signatures_against_maintainers()?;
+        self.verify_signatures()?;
+        Ok(())
+    }
+
+    pub fn is_valid(&self) -> bool {
+        match self.check_validity() {
+            Err(_) => false,
+            Ok(()) => true,
+        }
+    }
+
+    pub fn check_update(&self, previous: &Project) -> Result<(), UpdateVerificationError> {
+        if self.revision() <= previous.revision() {
+            return Err(UpdateVerificationError::NonMonotonicRevision);
+        }
+
+        let mut previous_maintainers = HashSet::<&PeerId>::new();
+        for m in previous.maintainers.iter() {
+            previous_maintainers.insert(m);
+        }
+        let retained = self
+            .maintainers
+            .iter()
+            .filter(|m| previous_maintainers.contains(m))
+            .count();
+        let total = self.maintainers.len();
+        let added = total - retained;
+        let removed = previous_maintainers.len() - retained;
+        let quorum = total / 2;
+
+        if added > quorum {
+            Err(UpdateVerificationError::NoCurrentQuorum)
+        } else if removed > quorum {
+            Err(UpdateVerificationError::NoPreviousQuorum)
+        } else {
+            Ok(())
+        }
+    }
+
+    pub fn check_history<'a>(
+        history: impl Iterator<Item = &'a Project>,
```

Comment:
I feel like this function might make more sense if `history` was `NonEmpty`. The inner logic would make more sense by doing:
```rust
let (current, previous) = history.split_first();
/* rest of code as normal */
```

And then the caller of _this_ function can attempt to turn the `history` into `NonEmpty`, returning `EmptyHistory` if there's a failure.

Reply:
I wanted to have an `Iterator` in the signature to avoid dictating a container implementation, and allow for eventual lazy loading of git commits and objects while verifying a history.
Ultimately I want to be able to verify a whole history without necessarily loading it all in RAM.

Reply:
However, you are dictating that the history must be non empty. Another way of doing it could be like:
```rust
 pub fn check_history<'a>(
      current: &'a Project,
      previous_history: impl Iterator<Item = &'a Project>,
)
```

But if you feel like this is the way to go, I'm not going to block you on this :)